docs/csharp/fundamentals/tutorials/file-based-programs.md

```diff
@@ -0,0 +1,246 @@
+
+That handles command line arguments correctly. Now, add the code to handle reading input from standard input (`stdin`) instead of command line arguments. Add the following `else` clause to the `if` statement you added in the preceding code:
+
+```csharp
+else
+{
+    while (Console.ReadLine() is string line && line.Length > 0)
+    {
+        Console.WriteLine(line);
+    }
+}
+```
+
+The preceding code reads the console input until either a blank line or a `null` is read. (The `ReadLine` method returns `null` if the input stream is closed by typing <kbd>ctrl+C</kbd>.) You can test reading standard input by creating a new text file in the same folder. Name the file `input.txt` and add the following lines:
+
+```text
+This is the input file
+for a file based program.
+It prints the messages
+from a file or the
+command line.
+There are options you
+can choose for ASCII
+art and colors.
+```
+
+Keep the lines short so they format correctly when you add the feature to use ASCII art. Now, run the program again:
+
+```dotnetcli
+cat input.txt | dotnet run AsciiArt.cs
+```
+
+Now your program can accept either command line arguments or standard input.
+
+> [!NOTE]
+>
+> Support for `#!` directives applies on unix platforms only. There isn't a similar directive for Windows to directly execute a C# program. On Windows, you must use `dotnet run` on the command line.
+
+On unix, you can run file-based programs directly, typing the source file name on the command line instead of `dotnet run`. You need to make two changes. First, set *execute* permissions on the source file:
+
+```bash
+chmod +x AsciiArt.cs
+```
+
+Then, add a `#!` directive as the first line of the `AsciiArt.cs` file:
+
+```csharp
+#!/usr/local/share/dotnet/dotnet run
+```
+
+The location of `dotnet` can be different on different unix installations. Use the command `whence dotnet` to local the `dotnet` host in your environment.
+
+After making these two changes, you can run the program from the command line directly:
+
+```bash
+./AsciiArt.cs
+```
+
+If you prefer, you can remove the extension so you can type `./AsciiArt` instead. You can add the `#!` to your source file even if you use Windows. The Windows command line doesn't support `#!`, but the C# compiler allows that directive in file based programs on all platforms.
+
+## Add features and NuGet packages to the program
+
+Next, add a package that supports ASCII art, [Colorful.Console](https://www.nuget.org/packages/Colorful.Console). To add a package to a file based program, you use the `#:package` directive. Add the following directive after the `#!` directive in your AsciiArt.cs file:
+
+:::code language="csharp" source="./snippets/file-based-programs/AsciiArt.cs" id="ColorfulPackage":::
+
+> [!IMPORTANT]
+>
+> The version `1.2.15` was the latest version when this tutorial was last updated. If there's a newer version available, use the latest version to ensure you have the latest security packages.
```

Comment:
Great point! It would be good to link to the package here (even though you did above) just because of the context. I think being direct is beneficial here too, especially for those that may not be familiar with not using an IDE to look up packages (since that probably doesn't work for file-based programs)


```suggestion
> The version `1.2.15` was the latest version of the `Colorful.Console` package when this tutorial was last updated. Check the package's [NuGet page](https://www.nuget.org/packages/Colorful.Console) for the latest version to ensure you use a package version with the latest security fixes.
```